kof-operator/internal/controller/configmap_controller.go

```diff
@@ -0,0 +1,576 @@
+			groupName := ruleGroup.Name
+
+			alertRules, ok := clusterGroupAlertRules[DefaultClusterName][groupName]
+			if !ok {
+				alertRules = AlertRules{}
+			}
+
+			recordRules, ok := clusterGroupRecordRules[DefaultClusterName][groupName]
+			if !ok {
+				recordRules = RecordRules{}
+			}
+
+			for _, rule := range ruleGroup.Rules {
+				if rule.Alert != "" {
+					// To avoid `field keep_firing_for not found in type rulefmt.RuleNode` in promxy:
+					rule.KeepFiringFor = nil
+					alertRules[rule.Alert] = rule
+				}
+				if rule.Record != "" {
+					recordRules = append(recordRules, rule)
+				}
+			}
+
+			if len(alertRules) > 0 {
+				clusterGroupAlertRules[DefaultClusterName][groupName] = alertRules
+			}
+
+			if len(recordRules) > 0 {
+				clusterGroupRecordRules[DefaultClusterName][groupName] = recordRules
+			}
+		}
+	}
+
+	return nil
+}
+
+// Merge the alert rules from `ConfigMaps` to the nested map.
+func mergeAlertConfigMaps(
+	ctx context.Context,
+	alertConfigMaps []corev1.ConfigMap,
+	clusterGroupAlertRules map[string]map[string]AlertRules,
+) error {
+	for _, configMap := range alertConfigMaps {
+		clusterName := configMap.Labels[KofAlertRulesClusterNameLabel]
+		groupAlertRules, ok := clusterGroupAlertRules[clusterName]
+		if !ok {
+			groupAlertRules = map[string]AlertRules{}
+			clusterGroupAlertRules[clusterName] = groupAlertRules
+		}
+		for groupName, alertRulesYAML := range configMap.Data {
+			alertRules, ok := groupAlertRules[groupName]
+			if !ok {
+				alertRules = AlertRules{}
+				groupAlertRules[groupName] = alertRules
+			}
+
+			newAlertRules := make(AlertRules)
+			if err := yaml.Unmarshal([]byte(alertRulesYAML), &newAlertRules); err != nil {
+				utils.LogEvent(
+					ctx,
+					"RulesUnmarshalFailed",
+					"Failed to unmarshal alert rules",
+					&configMap,
+					err,
+					"cluster", clusterName,
+					"group", groupName,
+					"rules", alertRulesYAML,
+				)
+				return err
+			}
+
+			for ruleName, newRule := range newAlertRules {
+				newRule.Alert = ruleName
+
+				oldRule, ok := alertRules[ruleName]
+				if ok {
+					// No need for deep copy here:
+					// default `ConfigMap should overwrite the data loaded from PrometheusRules,
+					// and cluster-specific ConfigMap will patch its own cluster rules only.
+					patchRule(&oldRule, &newRule)
+					alertRules[ruleName] = oldRule
+					continue
+				}
+
+				if clusterName != DefaultClusterName {
+					defaultRules, ok := clusterGroupAlertRules[DefaultClusterName][groupName]
+					if ok {
+						defaultRule, ok := defaultRules[ruleName]
+						if ok {
+							defaultRuleCopyPtr := defaultRule.DeepCopy()
+							patchRule(defaultRuleCopyPtr, &newRule)
+							alertRules[ruleName] = *defaultRuleCopyPtr
+							continue
+						}
+					}
+				}
+
+				alertRules[ruleName] = newRule
+			}
+		}
+	}
+
+	return nil
+}
+
+// Merge the record rules from `ConfigMaps` to the nested map.
+func mergeRecordConfigMaps(
```

Comment:
Probably it's better to move YAML decode to a separate function to simplify the main logic of this function